test-harness/tests/test_cache.py

```diff
@@ -637,3 +638,83 @@ def test_if_failed_request_triggered_update_is_recovered_by_timers(
 
             # Verify that the cache is OK now
             ping_mesos_agent(ar, valid_user_header)
+
+    def test_if_early_boot_stage_is_not_sensitive_to_failures(
+            self, nginx_class, valid_user_header, mocker, log_catcher):
+        # The idea here is to make Backend so slow, that AR is unable to
+        # update cache on first request. AR should then switch to fail-fast
+        # behaviour and wait for timers to update the backend when it recovers.
+
+        refresh_lock_timeout = 10
+        backend_request_timeout = 5
+
+        ar = nginx_class(cache_first_poll_delay=1,
+                         cache_poll_period=3,
+                         cache_expiration=2,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         cache_backend_request_timeout=backend_request_timeout,
+                         cache_refresh_lock_timeout=refresh_lock_timeout,
+                         )
+        agent_id = 'de1baf83-c36c-4d23-9cb0-f89f596cd6ab-S1'
+        url = ar.make_url_from_path('/agent/{}/blah/blah'.format(agent_id))
+        v = Vegeta(log_catcher, target=url, jwt=valid_user_header, rate=3)
+
+        # Make mesos slooowwwww.....
+        # It mus respond slower than backend_request_timeout
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_stall',
+                            aux_data=backend_request_timeout * 1.5)
+
+        with GuardedSubprocess(ar):
+            with GuardedSubprocess(v):
+                time.sleep(1)  # let it warm-up!
+
+                # Vegeta by this time should have already clogged AR
```

Comment:
> I did not want to write new code. Subprocess class is already there, Vegeta is highly configurable. Recreating constant-rate request functionality in Python reliably would take some time.

My issue is not with using Vegeta in general, obviously we should use it if we're benchmarking AR. My issue with the current approach is that we're not benchmarking anything here, nor should we be clogging any kind of sink. This just obscures the failure condition that we're testing here.

For example, let's say this test succeeds after some patching, it won't tell us anything about at what point the requests started failing fast, etc. Having a statically configured number of requests (ie. max worker connections) captures the essence of the bug perfectly without resorting to statistical reasoning (the equivalent of "just add a couple of sleeps".)

We don't need to perform requests at a constant-rate. All we need is a mock mesos endpoint that blocks indefinitely, and then make 1,024 requests to it, followed by a final request that we assert fails.